crates/cheatnet/src/cheatcodes/spy_events.rs

```diff
@@ -0,0 +1,47 @@
+use crate::CheatnetState;
+use cairo_felt::Felt252;
+use starknet_api::core::ContractAddress;
+
+pub struct Event {
+    pub from: ContractAddress,
+    pub name: Felt252,
+    pub keys: Vec<Felt252>,
+    pub data: Vec<Felt252>,
+}
+
+pub enum SpyOn {
+    All,
+    One(ContractAddress),
+    Multiple(Vec<ContractAddress>),
+}
+
+impl CheatnetState {
+    pub fn spy_events(&mut self, spy_on: SpyOn) {
+        self.cheatcode_state.spy_events = Some(spy_on);
+    }
+
+    pub fn fetch_events(&mut self) -> (usize, Vec<Felt252>) {
+        let serialized_events: Vec<Vec<Felt252>> = self
+            .cheatcode_state
+            .emitted_events
+            .iter()
+            .map(|event| {
+                let mut flattened: Vec<Felt252> = vec![
+                    Felt252::from_bytes_be(event.from.0.key().bytes()),
+                    event.name.clone(),
+                    Felt252::from(event.keys.len()),
+                ];
+                flattened.append(&mut event.keys.clone());
+                flattened.push(Felt252::from(event.data.len()));
+                flattened.append(&mut event.data.clone());
+
+                flattened
+            })
+            .collect();
+
+        let emitted_events_len = self.cheatcode_state.emitted_events.len();
+        self.cheatcode_state.emitted_events = vec![];
```

Comment:
Yup, we are removing them to avoid flattening a large number of events each time. They are stored in the cairo code under the `EventSpy.event` and we append them while calling `spy.fetch_events()`